File: fwScheduleBot/bot.py
import os
import discord
import logging
import requests
import operator
import json
import asyncio
from discord.ext import commands, tasks
from datetime import datetime, timedelta
from requests.api import head

__location__ = os.path.realpath(os.path.join(
    os.getcwd(), os.path.dirname(__file__)))
current_season = 25
current_teams = {0: "Unknown or undefined team"}
current_events = {}
current_players = {}
config = {
    "Tier1ScheduleChannel": 0,
    "Tier2ScheduleChannel": 0,
    "Tier3ScheduleChannel": 0,
    "Tier4ScheduleChannel": 0,
    "Tier5ScheduleChannel": 0,
    "DailyScheduleChannel": 0,
    "Tier1SeasonId": 0,
    "Tier2SeasonId": 0,
    "Tier3SeasonId": 0,
    "Tier4SeasonId": 0,
    "Tier5SeasonId": 0,
    "CurrentSeason": 25,
    "CurrentWeek": 0,
    "CurrentTeams": current_teams,
    "CurrentEvents": current_events,
    "CurrentPlayers": current_players,
}
authorized_roles = ['〈 👑 〉 Owner',
                    '〈 🛠️ 〉 League Director', '〈 👾 〉 League Commissioner']


def get_current_season_teams():
    resp = requests.get(
        "https://firewallesports.com/wp-json/sportspress/v2/teams?_fields=title,id&per_page=100&seasons={}".format(config['CurrentSeason']))

    teams = {0: "Unknown or undefined team"}

    for team in resp.json():
        teams[team['id']] = team['title']['rendered']

    return dict(sorted(teams.items(), key=operator.itemgetter(1)))


def get_all_events():
    event_loop_count = 0
    total_event_number = 1

    while event_loop_count < int(total_event_number):
        resp = requests.get(
            "https://firewallesports.com/wp-json/sportspress/v2/events?per_page=100&orderby=date&order=desc&offset={}".format(event_loop_count))
        for event in resp.json():
            current_events[event['id']] = event

        total_event_number = resp.headers['X-WP-Total']
        event_loop_count += 100

def get_all_players():
    player_loop_count = 0
    total_player_number = 1

    while player_loop_count < int(total_player_number):
        resp = requests.get(
            "https://firewallesports.com/wp-json/sportspress/v2/players?per_page=100&offset={}".format(player_loop_count))
        for event in resp.json():
            current_events[event['id']] = event

        total_player_number = resp.headers['X-WP-Total']
        player_loop_count += 100


def load_config():
    if (os.path.exists(__location__+"\\fwScheduleConfig.json")):
        print("Loading configuration from: {}".format(
            __location__+"\\fwScheduleConfig.json"))
        with open(__location__+"\\fwScheduleConfig.json", 'r') as openfile:
            global config
            config = json.load(openfile)


def save_config(config):
    with open(__location__+"\\fwScheduleConfig.json", "w") as outfile:
        json.dump(config, outfile)


def authorized_author(roles):
    for role in roles:
        if role.name in authorized_roles:
            return True
    return False


def refresh_data():
    global current_season, current_teams, current_events, config
    current_teams = get_current_season_teams()
    get_all_events()
    get_all_players()
    config["CurrentEvents"] = current_events
    config["CurrentTeams"] = current_teams
    config["CurrentPlayers"] = current_players
    save_config(config)

intents = discord.Intents.all()

bot = commands.Bot(command_prefix='??', description="Firewall Schedule Bot", intents=intents)

@bot.event
async def on_ready():
    print('Logged on as ', bot.user.name)
    save_config(config)
    activity = discord.Activity(
        name="??help", type=discord.ActivityType.playing)
    daily_refresh.start()
    await bot.change_presence(activity=activity)

async def report_week_tier(tier, week, newMessage=False):
    fields = []
    # put together a list of all the events for the week
    for event in current_events:
        if current_events[event]['day'] == week and int(current_events[event]['seasons'][0]) == int(config['CurrentSeason']) and int(current_events[event]['leagues'][0]) == int(config['Tier{}SeasonId'.format(tier)]):
            team1name = current_teams[current_events[event]['teams'][0]]
            team2name = current_teams[current_events[event]['teams'][1]]
            team1score = current_events[event]['results'][str(
                current_events[event]['teams'][0])]['mapscore']
            team2score = current_events[event]['results'][str(
                current_events[event]['teams'][1])]['mapscore']
            name = "{} VS {}".format(team1name, team2name)
            if team1score == "":
                event_time = datetime.strftime(datetime.fromisoformat(
                    current_events[event]['date']), "%A, %B %d, %I:%M %p")
                value = "{}: {} VS {}".format(
                    event_time, team1name, team2name)
            else:
                value = "Score: {}: {} VS {}: {}".format(
                    team1name, team1score, team2name, team2score)
            fields.append({"name": name, "value": value, })
    # if no events for the week, post a message saying so:
    if len(fields) == 0:
        fields.append({"name": "Week {}".format(week),
                       "value": "No games scheduled for Week {}".format(week)})
    footer = {
        'text': 'Generated by fwScheduleBot - Last Updated: {}'.format(datetime.now())}
    data = {"title": "Tier {} Week {} Schedule".format(tier, week),
            "fields": fields,
            "footer": footer, }
    embed = discord.Embed.from_dict(data)
    message_configuration = 'Tier{}ScheduleMessageWeek{}Season{}'.format(
        tier, week, current_season)
    channel = bot.get_channel(
        config['Tier{}ScheduleChannel'.format(tier)])
    if message_configuration in config:
        message = await channel.fetch_message(config[message_configuration])
        await message.edit(embed=embed)
    else:
        message = await channel.send(content=None, embed=embed)
        config[message_configuration] = message.id
        save_config(config)
# go through all of the current events, find all that have a date today, and post them to the channel
async def report_today_event(self, newMessage=False):
    if config['DailyScheduleChannel'] == 0:
        logger.warn("Daily schedule channel not set in config, skipping daily schedule posting")
        return
    fields = []
    for event in current_events:
        event_date = datetime.strftime(datetime.fromisoformat(
            current_events[event]['date']), "%x")
        if event_date == datetime.today().strftime("%x"):
            team1name = current_teams[current_events[event]['teams'][0]]
            team2name = current_teams[current_events[event]['teams'][1]]
            event_time = datetime.strftime(datetime.fromisoformat(
                current_events[event]['date']), "%I:%M %p")
            value = "{}: {} VS {}".format(event_time, team1name, team2name)
            fields.append({"name": "{} VS {}".format(
                team1name, team2name), "value": value})
    if len(fields) == 0:
        fields.append({"name": "No games scheduled for today.",
                      "value": "No games scheduled for today."})
    footer = {
        'text': 'Generated by fwScheduleBot - Last Updated: {}'.format(datetime.now())}
    data = {"title": "Today's Scheduled Games: {}".format(
        datetime.today().strftime("%A, %B %d")), "fields": fields, "footer": footer, }
    embed = discord.Embed.from_dict(data)
    message_configuration = 'TodayEventMessageId'
    channel = bot.get_channel(config['DailyScheduleChannel'])
    if message_configuration in config:
        message = await channel.fetch_message(config[message_configuration])
        await message.edit(embed=embed)
    else:
        message = await channel.send(content=None, embed=embed)
        config[message_configuration] = message.id
        save_config(config)

async def success_reaction(ctx):
    await ctx.message.add_reaction('✔')

async def error_reaction(ctx):
    await ctx.message.add_reaction('❌')

async def before_reaction(ctx):
    await ctx.message.add_reaction('⏳')

@bot.on_command_error
async def on_command_error(ctx, exception):
    if isinstance(exception, commands.CheckFailure):
        await ctx.channel.send("You don't have permission to use that command")
    elif isinstance(exception, commands.MissingRequiredArgument):
        await ctx.channel.send("Missing argument")
    elif isinstance(exception, commands.BadArgument):
        await ctx.channel.send("Bad Argument")
    elif isinstance(exception, commands.CommandInvokeError):
        await ctx.channel.send("Command Error")
    elif isinstance(exception, commands.CommandOnCooldown):
        await ctx.channel.send("Command on cooldown")
    else:
        await ctx.channel.send("Unknown Error")
    error_reaction(ctx)

@bot.command()
async def ping(ctx):
    await ctx.channel.send('pong')
    
@bot.command()
@commands.has_any_role('〈 👑 〉 Owner', '〈 🛠️ 〉 League Director', '〈 👾 〉 League Commissioner')
@commands.after_invoke(success_reaction)
@commands.before_invoke(before_reaction)
async def currentTeams(ctx):
    teams = config['CurrentTeams']
    output = ""
    for team in teams:
        output += "\r\n{}".format(teams[team])
    await ctx.channel.send(output)

@currentTeams.error
async def currentTeams_error(ctx, error):
    error_output(ctx, error)

@bot.command()
async def setScheduleChannel(ctx, tier: int, channel: discord.TextChannel):
    output = "Setting Tier {} channel to id {}".format(
        tier, channel)
    config['Tier{}ScheduleChannel'.format(tier)] = channel.id
    save_config(config)
    await ctx.channel.send(output)
    await ctx.message.add_reaction('✔')

@bot.command()
async def setTierId(ctx, tier: int, id: int):
    output = "Setting configuration value {} to {}".format(
        'Tier{}SeasonId'.format(tier), id)
    config['Tier{}SeasonId'.format(tier)] = id
    save_config(config)
    await ctx.channel.send(output)
    await ctx.message.add_reaction('✔')

@bot.command()
async def reportScheduleConfiguration(ctx):
    try:
        output = "Channels:\r\nTier 1: {}\r\nTier 2: {}\r\nTier 3: {}\r\nTier 4: {}\r\nTier 5: {}\r\nDaily: {}".format(
            config['Tier1ScheduleChannel'], config['Tier2ScheduleChannel'], config['Tier3ScheduleChannel'], config['Tier4ScheduleChannel'], config['Tier5ScheduleChannel'], config['DailyScheduleChannel'])
        output += "\r\nLeague Ids:\r\nTier 1: {}\r\nTier 2: {}\r\nTier 3: {}\r\nTier 4: {}\r\nTier 5: {}".format(
            config['Tier1SeasonId'], config['Tier2SeasonId'], config['Tier3SeasonId'], config['Tier4SeasonId'], config['Tier5SeasonId'])
        output += "\r\nCurrent Season: {}".format(current_season)
        output += "\r\nCurrent Week: {}".format(config['CurrentWeek'])
        output += "\r\nCurrent User: {}".format(ctx.author.roles)
    except Exception as e: 
        logger.error(e)
        output = f"Error getting configuration data: {e}."
    data = {"title": "Schedule Configuration",
            "fields": [{"name": "Configuration",
                        "value": output, }]
            }
    embed = discord.Embed.from_dict(data)
    await ctx.channel.send(content=None, embed=embed)
    await ctx.message.add_reaction('✔')

@bot.command()
async def reportWeekGames(ctx, week: int, newMessage: bool = False):
    tier = 1
    while tier < 5:
        await report_week_tier(tier, week, newMessage)
        tier += 1
    if newMessage:
        output = "Sent schedules for week {}".format(week)
    else:
        output = "Updated schedules for week {}".format(week)
    config['CurrentWeek'] = week
    save_config(config)
    await ctx.channel.send(output)
    await ctx.message.add_reaction('✔')

@bot.command()
async def reportDayGames(ctx, newMessage: bool = False):
    await report_today_event(newMessage)
    output = "Updated games for today"
    await ctx.channel.send(output)
    await ctx.message.add_reaction('✔')

@bot.command()
async def setCurrentSeason(ctx, season: int):
    config['CurrentSeason'] = season
    save_config(config)
    await ctx.channel.send("Set current season to {}".format(season))
    await ctx.message.add_reaction('✔')

@bot.command()
async def setDailyScheduleChannel(ctx, channel: discord.TextChannel):
    config['DailyScheduleChannel'] = channel.id
    save_config(config)
    await ctx.channel.send("Set daily schedule channel to {}".format(channel))
    await ctx.message.add_reaction('✔')

@bot.command()
async def resetTeams(ctx):
    current_teams = {0: "Unknown or undefined team"}
    current_teams = get_current_season_teams()
    config['CurrentTeams'] = current_teams
    save_config(config)
    await ctx.channel.send("Reset current teams")
    await ctx.message.add_reaction('✔')

@tasks.loop(hours=24)
async def daily_refresh(self):
    refresh_data()
    tier = 1
    while tier < 5:
        await report_week_tier(tier, config['CurrentWeek'], False)
        tier += 1
    await report_today_event(False)
    logging.info("Daily refresh complete at {}".format(datetime.now()))
    print("Testing the daily refresh complete at {}".format(datetime.now()))

@daily_refresh.before_loop
async def before_daily_refresh(self):
    await self.wait_until_ready()
    hour = 1
    minute = 0
    now = datetime.now()
    future = datetime(year=now.year, month=now.month, day=now.day, hour=hour, minute=minute)
    if now.hour >= hour and now.minute > minute:
        future += timedelta(days=1)
    await asyncio.sleep((future-now).seconds)


logger = logging.getLogger('discord')
logger.setLevel(logging.INFO)
handler = logging.FileHandler(
    filename=__location__ + '\\discord.log', encoding='utf-8', mode='w')
handler.setFormatter(logging.Formatter(
    '%(asctime)s:%(levelname)s:%(name)s: %(message)s'))
logger.addHandler(handler)

load_config()
refresh_data()
token = os.environ.get("BOT_TOKEN")
bot.run(token)
# https://discord.com/oauth2/authorize?client_id=858528014500757504&permissions=2684447824&scope=bot

File: fwScheduleBot/bot.ts
import { Client, Message, MessageEmbed, TextChannel } from "discord.js";
import fs = require("fs");
const axios = require("axios").default;
import dateFormat = require("dateformat");
import { Team } from "./models/team";
import { FWEvent } from "./models/FWEvent";
import { Player } from "./models/player";

class FirewallBot {
  private static _instance: FirewallBot;
  public config: ScheduleBotConfig;
  static _client: Client;

  constructor() {
    this.config = {
      Tier1PlayerMax: 2700,
      Tier2PlayerMax: 3200,
      Tier3PlayerMax: 3700,
      Tier4PlayerMax: 4250,
      Tier5PlayerMax: 9999,
      Tier1TeamMax: 2550,
      Tier2TeamMax: 3000,
      Tier3TeamMax: 3550,
      Tier4TeamMax: 4000,
      Tier5TeamMax: 4250,
      Tier1ResultsChannel: "0",
      Tier2ResultsChannel: "0",
      Tier3ResultsChannel: "0",
      Tier4ResultsChannel: "0",
      Tier5ResultsChannel: "0",
      ListCache: [] as Team[],
      MaxWeeks: 11,
    };
    this.config = Object.assign(
      this.config,
      JSON.parse(fs.readFileSync("./fwScheduleconfig.json", "utf8"))
    );
    this.get_current_season_teams();
    this.get_all_events();
    this.get_all_players();
  }

  public static Instance(client?: Client) {
    if (!FirewallBot._instance) {
      this._instance = new FirewallBot();
    }
    if (client as Client) this._client = client;
    return this._instance || (this._instance = new this());
  }

  saveConfig = () => {
    fs.writeFileSync("./fwScheduleconfig.json", JSON.stringify(this.config));
  };

  async get_current_season_teams() {
    const url = `https://firewallesports.com/wp-json/sportspress/v2/teams?_fields=title,id&per_page=100&seasons=${this.config.CurrentSeason}`;
    try {
      const response = await axios.get(url);
      const teams = [{ id: 0, name: "Unknown or undefined team" }] as Team[];
      for (const team of response.data) {
        teams.push({ id: team.id, name: team.title.rendered });
      }

      this.config.CurrentTeams = teams;
      this.saveConfig();
    } catch (error) {
      console.log(error);
    }
  }

  async get_all_events() {
    var event_loop_count = 0;
    var total_event_number = 1;
    const current_events = [] as FWEvent[];

    while (event_loop_count < total_event_number) {
      try {
        const response = await axios.get(
          `https://firewallesports.com/wp-json/sportspress/v2/events?per_page=100&orderby=date&order=desc&offset=${event_loop_count}`
        );
        for (const event of response.data) {
          current_events.push(event);
        }

        total_event_number = response.headers["x-wp-total"];
        event_loop_count += 100;
      } catch (error) {
        console.log(error);
      }
    }
    this.config.CurrentEvents = current_events;
    this.saveConfig();
  }

  async get_all_players() {
    var player_loop_count = 0;
    var total_player_number = 1;
    const current_players = [] as Player[];

    while (player_loop_count < total_player_number) {
      try {
        const response = await axios.get(
          `https://firewallesports.com/wp-json/sportspress/v2/players?per_page=100&orderby=date&order=desc&offset=${player_loop_count}`
        );
        for (const player of response.data) {
          current_players.push(player);
        }
        total_player_number = response.headers["x-wp-total"];
        player_loop_count += 100;
      } catch (error) {
        console.log(error);
      }
    }
    this.config.CurrentPlayers = current_players;
    this.saveConfig();
  }

  async refresh_data() {
    await this.get_all_events();
    await this.get_all_players();
    await this.get_current_season_teams();
  }

  async get_player_list_by_id(id: number): Promise<Player[]> {
    const players = [];
    try {
      const response = await axios.get(
        `https://firewallesports.com/wp-json/sportspress/v2/lists/${id}`
      );
      for (const player of Object.values(response.data.data) as Player[]) {
        player.sr = Math.max(
          player.dpssr,
          player.dpssrs,
          player.tanksr,
          player.tanksrtwo,
          player.supportsr,
          player.supportsrtwo
        );
        if (Number.isInteger(player.sr)) players.push(player);
      }
      this.cache_list_name(id, response.data.title.rendered);
    } catch (error) {
      console.log(error);
    }
    return players;
  }

  async get_srs_from_player_list(list: Player[]): Promise<number[]> {
    const srs = [];
    for (const player of list) {
      const playersr = [
        player.dpssr,
        player.dpssrs,
        player.tanksr,
        player.tanksrtwo,
        player.supportsr,
        player.supportsrtwo,
      ] as number[];
      const sr = Math.max(...playersr);
      if (Number.isInteger(sr)) srs.push(sr);
    }
    return srs;
  }

  async get_team_sr_average(team_players: Player[]): Promise<number> {
    const team_srs = await this.get_srs_from_player_list(team_players);
    team_srs.sort((a, b) => b - a);
    while (team_srs.indexOf(0) !== -1) team_srs.splice(team_srs.indexOf(0));
    const team_srs_for_average = team_srs.slice(0, 7);
    const team_sr_average =
      team_srs_for_average.reduce((a, b) => a + b, 0) /
      team_srs_for_average.length;
    return team_sr_average;
  }

  get_tier_team_max_sr(tier: number) {
    return this.config[`Tier${tier}TeamMax`];
  }

  get_tier_player_max_sr(tier: number) {
    return this.config[`Tier${tier}PlayerMax`];
  }

  get_team_name_by_id(id: number) {
    const team =
      this.config.CurrentTeams.find((t) => t.id === id) ||
      this.config.ListCache.find((t) => t.id === id);
    return team ? team.name : "Unknown or undefined team";
  }

  cache_list_name(id: number, name: string) {
    if (!this.config.ListCache.find((t) => t.id === id)) {
      this.config.ListCache.push({ id: id, name: name });
      this.saveConfig();
    } else {
      this.config.ListCache.find((t) => t.id === id).name = name;
    }
  }

  async reportDayGames(newMessage: boolean = false) {
    if (!this.config.DailyScheduleChannel) {
      console.log("Daily schedule channel not set");
      return false;
    }

    const current_events = this.config.CurrentEvents;
    const day_games = current_events.filter((e) =>
      this.are_dates_same_day(e.date, new Date())
    );

    const day_sorted_games = day_games.sort((a, b) => {
      if (a.date < b.date) {
        return -1;
      }
      if (a.date > b.date) {
        return 1;
      }
      return 0;
    });

    const embed = new MessageEmbed()
      .setTitle(`Today's Games: ${new Date().toDateString()}`)
      .setTimestamp()
      .setFooter(`Firewall Season 5`);

    for (const game of day_sorted_games) {
      const game_embed = this.build_event_output_embed(game as FWEvent);
      embed.addField(game_embed.fields[0].name, game_embed.fields[0].value);
    }

    if (day_sorted_games.length === 0) {
      embed.addField("No games today", "No games today");
    }

    console.log(
      "Daily Schedule Channel Id: " + this.config.DailyScheduleChannel
    );
    const channel = FirewallBot._client.channels.cache.get(
      this.config.DailyScheduleChannel
    ) as TextChannel;
    if (!channel) {
      console.log("Channel not found");
      return false;
    }

    const message = (await this.editMessageOrCreate(
      channel,
      this.config.DailyScheduleMessage,
      embed
    )) as Message;

    if (message) {
      this.config.DailyScheduleMessage = message.id;
    } else return false;

    return true;
  }

  are_dates_same_day(date1: Date, date2: Date) {
    return dateFormat(date1, "yyyy-mm-dd") === dateFormat(date2, "yyyy-mm-dd");
  }

  build_event_output_embed = (event: FWEvent) => {
    const embed = new MessageEmbed();

    const teams = event.teams.map((t) => this.get_team_name_by_id(t));

    if (event.main_results.length === 0) {
      embed.addField(
        teams.join(" VS "),
        `${dateFormat(event.date, "dddd mmm dd':' h TT Z")}: ${teams.join(
          " VS "
        )}`
      );
    } else {
      this.buildEventScoreEmbed(event, embed);
    }
    return embed;
  };

  buildEventScoreEmbed(event: FWEvent, embed: MessageEmbed) {
    const teams = event.teams.map((t) => this.get_team_name_by_id(t));
    if (!event.winner) return;
    const winnerIndex = event.teams.indexOf(event.winner);
    //index is always 0 or 1, so we can check falsy values to see which score to put where
    if (winnerIndex) {
      embed.addField(
        teams.join(" VS "),
        `Winner: ${this.get_team_name_by_id(event.winner)}: ${
          event.main_results[1]
        } - ${event.main_results[0]}`
      );
    } else {
      embed.addField(
        teams.join(" VS "),
        `Winner: ${this.get_team_name_by_id(event.winner)}: ${
          event.main_results[0]
        } - ${event.main_results[1]}`
      );
    }
  }

  async reportWeekGames(
    tier: number,
    week: number,
    newMessage: boolean = false
  ) {
    const scheduleChannel = this.config[`Tier${tier}ScheduleChannel`];
    const resultsChannel = this.config[`Tier${tier}ResultsChannel`];
    const tierId = this.config[`Tier${tier}SeasonId`];

    if (!scheduleChannel) {
      console.warn(`Schedule channel tier ${tier} not set - failing`);
      return false;
    }
    if (!resultsChannel) {
      console.warn(`Results channel tier ${tier} not set - failing`);
      return false;
    }

    const current_events = this.config.CurrentEvents;

    const week_games = current_events.filter(
      (e) =>
        e.day === week.toString() && e.leagues.some((r) => tierId.includes(r))
    );

    const week_sorted_games = week_games.sort((a, b) => {
      if (a.date < b.date) {
        return -1;
      }
      if (a.date > b.date) {
        return 1;
      }
      return 0;
    });

    const scheduleEmbed = new MessageEmbed()
      .setTitle(`Week ${week} Games: Tier ${tier}`)
      .setTimestamp()
      .setFooter(`Firewall Season 5`);

    const resultEmbed = new MessageEmbed()
      .setTitle(`Week ${week} Results: Tier ${tier}`)
      .setTimestamp()
      .setFooter(`Firewall Season 5`);

    for (const game of week_sorted_games) {
      const game_embed = this.build_event_output_embed(game as FWEvent);
      scheduleEmbed.addField(
        game_embed.fields[0].name,
        game_embed.fields[0].value
      );
      if (game.winner) {
        this.buildEventScoreEmbed(game, resultEmbed);
      }
    }

    if (week_games.length === 0) {
      scheduleEmbed.addField("No games this week", "No games this week");
    }

    const channel = FirewallBot._client.channels.cache.get(
      scheduleChannel
    ) as TextChannel;
    if (!channel) {
      console.log("Schedule channel not found");
      return false;
    }

    const scheduleMessageConfigName = `Tier${tier}Week${week}ScheduleMessage`;
    let messageId = this.config[scheduleMessageConfigName];
    if (newMessage) messageId = undefined;
    const message = (await this.editMessageOrCreate(
      channel,
      messageId,
      scheduleEmbed
    )) as Message;

    if (message) {
      this.config[scheduleMessageConfigName] = message.id;
    } else return false;

    const resultChannel = FirewallBot._client.channels.cache.get(
      resultsChannel
    ) as TextChannel;
    if (!resultChannel) {
      console.log("Results channel not found");
      return false;
    }

    const resultMessageConfigName = `Tier${tier}Week${week}ResultsMessage`;
    messageId = this.config[resultMessageConfigName];
    if (newMessage) messageId = undefined;
    if (resultEmbed.fields.length > 0) {
      const resultMessage = (await this.editMessageOrCreate(
        resultChannel,
        messageId,
        resultEmbed
      )) as Message;

      if (resultMessage) {
        this.config[resultMessageConfigName] = resultMessage.id;
      } else return false;
    }

    const casterConfigName = `Tier${tier}Week${week}CasterScheduleMessage`;
    messageId = this.config[casterConfigName];
    if (newMessage) messageId = undefined;
    const casterChannel = FirewallBot._client.channels.cache.get(
      this.config[`CasterScheduleChannel`]
    ) as TextChannel;
    if (!casterChannel) {
      console.warn("Caster channel not found");
    } else {
      const casterMessage = (await this.editMessageOrCreate(
        casterChannel,
        messageId,
        scheduleEmbed
      )) as Message;

      if (casterMessage) {
        this.config[casterConfigName] = casterMessage.id;
      } else {
        console.warn(`Caster message not sent - Tier ${tier}`);
        return false;
      }
    }

    this.saveConfig();

    return true;
  }

  async editMessageOrCreate(
    channel: TextChannel,
    messageId: string,
    embed: MessageEmbed
  ) {
    if (!messageId) {
      console.log("Creating new message");
      try {
        const message = await channel.send({ embeds: [embed] });
        return message;
      } catch (error) {
        console.error(error);
        return false;
      }
    } else {
      const message = await channel.messages.fetch(messageId);
      try {
        await message.edit({ embeds: [embed] });
      } catch (error) {
        console.error(error);
        try {
          console.warn("Failed to find message - resending");
          const message = await channel.send({ embeds: [embed] });
          return message;
        } catch (error) {
          console.error(error);
          return false;
        }
      }
      return message;
    }
  }

  async reportAllWeekGames(currentWeek?: number, newMessage: boolean = false) {
    if (!currentWeek) {
      currentWeek = this.config.CurrentWeek;
    }

    for (let week = currentWeek; week > 0; week--) {
      for (let i = 1; i <= 5; i++) {
        try {
          await this.reportWeekGames(i, week, newMessage);
        } catch (error) {
          console.error(error.stack);
          return false;
        }
      }
    }
    if (currentWeek !== this.config.MaxWeeks) {
      for (let i = 1; i <= 5; i++) {
        try {
          await this.reportWeekGames(i, currentWeek + 1, newMessage);
        } catch (error) {
          console.error(error.stack);
          return false;
        }
      }
    }
    return true;
  }
}

interface ScheduleBotConfig extends Record<string, any> {
  Tier1ScheduleChannel?: string;
  Tier2ScheduleChannel?: string;
  Tier3ScheduleChannel?: string;
  Tier4ScheduleChannel?: string;
  Tier5ScheduleChannel?: string;
  DailyScheduleChannel?: string;
  DailyScheduleMessage?: string;
  CasterScheduleChannel?: string;
  Tier1ResultsChannel?: string;
  Tier2ResultsChannel?: string;
  Tier3ResultsChannel?: string;
  Tier4ResultsChannel?: string;
  Tier5ResultsChannel?: string;
  Tier1SeasonId?: number[];
  Tier2SeasonId?: number[];
  Tier3SeasonId?: number[];
  Tier4SeasonId?: number[];
  Tier5SeasonId?: number[];
  Tier1PlayerMax?: number;
  Tier2PlayerMax?: number;
  Tier3PlayerMax?: number;
  Tier4PlayerMax?: number;
  Tier5PlayerMax?: number;
  Tier1TeamMax?: number;
  Tier2TeamMax?: number;
  Tier3TeamMax?: number;
  Tier4TeamMax?: number;
  Tier5TeamMax?: number;
  CurrentSeason?: number;
  CurrentWeek?: number;
  MaxWeeks?: number;
  CurrentTeams?: Team[];
  CurrentEvents?: FWEvent[];
  CurrentPlayers?: Player[];
  ListCache?: Team[];
}

export { FirewallBot };
